app/assets/javascripts/discourse/routes/unknown.js

```diff
@@ -1,8 +1,28 @@
-import DiscourseRoute from "discourse/routes/discourse";
 import { ajax } from "discourse/lib/ajax";
+import DiscourseURL from "discourse/lib/url";
+import DiscourseRoute from "discourse/routes/discourse";
+import RSVP from "rsvp";
 
 export default DiscourseRoute.extend({
-  model() {
-    return ajax("/404-body", { dataType: "html" });
+  model(params, transition) {
+    const path = params.path;
+    const bodyReq = ajax("/404-body", { dataType: "html" });
+    const testReq = ajax("/permalink-check.json", {
+      data: { path }
+    });
+    return RSVP.allSettled([bodyReq, testReq]).then(results => {
+      const [bodyResp, testResp] = results;
+      if (testResp.state === "fulfilled" && testResp.value) {
```

Comment:
Checking promise response looks a bit hacky. Also, it is always doing two requests even if it should not be the case, but I understand the intention of having parallel requests. 

Is there any chance you could make the `/permalink-check` endpoint return the 404-body if permalink does not exist?

```javascript
ajax("/permalink-check", { data: { path }, dataType: "html" })
  .then(url => { /* redirect */ })
  .catch(body404 => { /* show error */ });
```

Reply:
Agreed, I don't think we should count on `state`/`value` for this.

Reply:
Ported the code from exceptions#not_found_body into the check_permalink handler. This looks even cleaner on the client now!